2020/06/02/why-upgrading-software-libraries-is-imperative.html.md

```diff
@@ -0,0 +1,44 @@
+---
+author: "Selvakumar Arumugam"
+title: "Why Upgrading Software Libraries is Imperative"
+tags: software, version, upgrade
+gh_issue_number: 
+---
+
+Stack applications primarily consist of the Operating System as well as front and backend programming languages including their necessary list of library packages. The Operating System and programming language software can be periodically updated and run on the latest version. However, what about the long list of libraries being used in the application’s front and backend? As we know, It can be quite daunting to maintain and individually update a long list of software libraries. Still, it is important to keep them updated in several ways while mitigating the level of complexity. This post dives into our experience upgrading a complex stack with a list of libraries. What you will need, how to go about it, and the benefits of such an activity are all examined below. 
+
+The application stack in question contained decade-old software and included extensive libraries when we received it from our client. The application consisted of languages including Java, Scala, Kotlin and JavaScript in tandem with an extensive list of libraries. The initial plan was to upgrade the complete stack all at once due to the gap between versions. This proved to be more difficult than expected due to a host of deprecated and removed functionalities in addition to their inter-dependency on a few of the libraries. 
+
+### Conflict Approach
+Speaking on upgrading libraries and dependencies,it is often said: “don't update unless you have to.” For the short term, this can be sustainable, however, not as much in the long run. Primarily, the updates protect from known vulnerabilities and cyber attacks. Scenarios arise where particular library fixes are implemented on the latest version, yet require upgrading other libraries to the latest version in a chain. Because an upgraded list of libraries needs complete testing and preparation for new issues, this directly impacts whether the application attempts to resolve an issue. Therefore, smaller and more frequent updates are more sustainable in the long run. Larger and more infrequent upgrades will not only result in unexpected errors, but also require regress testing to deliver a version free of error. The following points justify the necessity of software and library upgrades:
+
+* Apply timely security patches to reduce vulnerabilities and defend from cyberattack
+* Identify deprecated functionality earlier and use alternatives
+* Remain on your Long Term Support (LTS) version
```

Comment:
Note that not everything has an LTS version -- especially many libraries do not.

It would probably be good to explain what the difference is between a minor update or patch release that isn't supposed to break compatibility and can be applied fairly safely without big cost or work, vs. a major update or upgrade that often does break compatibility and needs extensive testing and/or rewriting.

Reply:
@jonjensen Added details on major, minor and patch release